File: Data Structures & Algorithms/max-water-container/submission-5.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        best = 0 
        left = 0
        right = len(heights) - 1

        while left < right:
            best = max(best,min(heights[left], heights[right]) * (right - left))
            
            if heights[left] < heights[right]:
                left += 1
            else:
                right -= 1 
    
        return best
            
            
        

File: Data Structures & Algorithms/eating-bananas/submission-5.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int left = 1;
        int right = *max_element(piles.begin(), piles.end());        
        int best = right;

        while (left <= right) {
            int hours = 0;
            int mid = (right + left) / 2;

            for(int pile: piles) {
                hours += (pile + mid - 1) / mid;
            } 

            if(hours > h) {
                left = mid + 1;
            } else {
                best = min(best,mid);
                right = mid - 1;
            }
        }
        
        return best;
    }
};
